File: 2_ReactWeb/src/services/system/announcements.ts
import { env } from "../../shared/config/env";
import { fetchJson } from "../http/httpClient";

export type AnnouncementImportance = "normal" | "important" | "critical";
export type AnnouncementStatus = "published" | "withdrawn";

export type AnnouncementYearReference = {
  year: number;
  indexPath: string;
};

export type AnnouncementRootIndex = {
  schemaVersion: number;
  updatedAt: string;
  latestAnnouncementId: string;
  latestAnnouncementYear: number;
  years: AnnouncementYearReference[];
};

export type AnnouncementItem = {
  id: string;
  revision: number;
  title: string;
  summary: string;
  publishedAt: string;
  importance: AnnouncementImportance;
  status: AnnouncementStatus;
  contentPath: string;
  read: boolean;
};

export type AnnouncementYearIndex = {
  schemaVersion: number;
  year: number;
  updatedAt: string;
  announcements: AnnouncementItem[];
  cached: boolean;
};

export type AnnouncementSettings = {
  source: string;
  checkOnStartup: boolean;
};

export type AnnouncementCheck = {
  root: AnnouncementRootIndex;
  latestYear: AnnouncementYearIndex;
  latest: AnnouncementItem;
  latestUnread: boolean;
  cached: boolean;
  lastSuccessfulCheckAt: string | null;
};

export type AnnouncementContent = {
  announcement: AnnouncementItem;
  content: string;
  cached: boolean;
};

export function getAnnouncementSettings(signal?: AbortSignal) {
  return fetchJson<AnnouncementSettings>("/api/announcements/settings", { signal });
}

export function updateAnnouncementSettings(checkOnStartup: boolean) {
  return fetchJson<AnnouncementSettings>("/api/announcements/settings", {
    method: "PUT",
    body: JSON.stringify({ checkOnStartup }),
  });
}

let checkInFlight: Promise<AnnouncementCheck> | null = null;

export function checkAnnouncements() {
  checkInFlight ??= fetchJson<AnnouncementCheck>("/api/announcements/check", {
    method: "POST",
  }).finally(() => {
    checkInFlight = null;
  });
  return checkInFlight;
}

export function getAnnouncementYear(year: number, signal?: AbortSignal) {
  return fetchJson<AnnouncementYearIndex>(`/api/announcements/years/${year}`, { signal });
}

export function getAnnouncementContent(
  announcementId: string,
  revision: number,
  signal?: AbortSignal,
) {
  return fetchJson<AnnouncementContent>(
    `/api/announcements/${encodeURIComponent(announcementId)}/content?revision=${revision}`,
    { signal },
  );
}

export function markAnnouncementRead(announcementId: string, revision: number) {
  return fetchJson<{ announcementId: string; revision: number; readAt: string }>(
    `/api/announcements/${encodeURIComponent(announcementId)}/read`,
    { method: "POST", body: JSON.stringify({ revision }) },
  );
}

export function resolveAnnouncementAssetUrl(
  announcement: Pick<AnnouncementItem, "id" | "revision">,
  rawSource: string | undefined,
) {
  if (!rawSource) return "";
  const normalized = rawSource.replace(/\\/g, "/");
  if (
    !normalized.startsWith("assets/")
    || normalized.split("/").some((part) => part === "" || part === "." || part === "..")
  ) {
    return "";
  }
  const path = normalized
    .slice("assets/".length)
    .split("/")
    .map(encodeURIComponent)
    .join("/");
  return `${env.apiBaseUrl}/api/announcements/${encodeURIComponent(announcement.id)}/assets/${path}?revision=${announcement.revision}`;
}

File: 1_PythonServer/app/schemas/announcements.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class _AnnouncementModel(BaseModel):
    model_config = ConfigDict(populate_by_name=True)


class AnnouncementYearReference(_AnnouncementModel):
    year: int
    index_path: str = Field(alias="indexPath")


class AnnouncementRootIndex(_AnnouncementModel):
    schema_version: int = Field(alias="schemaVersion")
    updated_at: str = Field(alias="updatedAt")
    latest_announcement_id: str = Field(alias="latestAnnouncementId")
    latest_announcement_year: int = Field(alias="latestAnnouncementYear")
    years: list[AnnouncementYearReference]


class AnnouncementItem(_AnnouncementModel):
    id: str
    revision: int
    title: str
    summary: str
    published_at: str = Field(alias="publishedAt")
    importance: Literal["normal", "important", "critical"]
    status: Literal["published", "withdrawn"]
    content_path: str = Field(alias="contentPath")
    read: bool = False


class AnnouncementYearIndex(_AnnouncementModel):
    schema_version: int = Field(alias="schemaVersion")
    year: int
    updated_at: str = Field(alias="updatedAt")
    announcements: list[AnnouncementItem]
    cached: bool = False


class AnnouncementSettings(_AnnouncementModel):
    source: str
    check_on_startup: bool = Field(default=True, alias="checkOnStartup")


class AnnouncementSettingsUpdate(_AnnouncementModel):
    check_on_startup: bool = Field(alias="checkOnStartup")


class AnnouncementCheckResponse(_AnnouncementModel):
    root: AnnouncementRootIndex
    latest_year: AnnouncementYearIndex = Field(alias="latestYear")
    latest: AnnouncementItem
    latest_unread: bool = Field(alias="latestUnread")
    cached: bool
    last_successful_check_at: str | None = Field(alias="lastSuccessfulCheckAt")


class AnnouncementContentResponse(_AnnouncementModel):
    announcement: AnnouncementItem
    content: str
    cached: bool


class AnnouncementReadRequest(_AnnouncementModel):
    revision: int


class AnnouncementReadResponse(_AnnouncementModel):
    announcement_id: str = Field(alias="announcementId")
    revision: int
    read_at: str = Field(alias="readAt")


class AnnouncementAssetReference(_AnnouncementModel):
    path: str
    media_type: str | None = Field(default=None, alias="mediaType")
